// icodec/lib/avif.js
import wasmFactoryEnc from "../dist/avif-enc.js";
import wasmFactoryDec from "../dist/avif-dec.js";
import { check, encodeES, loadES } from "./common.js";
export var Subsampling;
(function (Subsampling) {
    Subsampling[Subsampling["YUV444"] = 1] = "YUV444";
    Subsampling[Subsampling["YUV422"] = 2] = "YUV422";
    Subsampling[Subsampling["YUV420"] = 3] = "YUV420";
    Subsampling[Subsampling["YUV400"] = 4] = "YUV400";
})(Subsampling || (Subsampling = {}));
export var AVIFTune;
(function (AVIFTune) {
    AVIFTune[AVIFTune["Auto"] = 0] = "Auto";
    AVIFTune[AVIFTune["PSNR"] = 1] = "PSNR";
    AVIFTune[AVIFTune["SSIM"] = 2] = "SSIM";
})(AVIFTune || (AVIFTune = {}));
export const defaultOptions = {
    quality: 50,
    qualityAlpha: -1,
    speed: 6,
    subsample: Subsampling.YUV420,
    autoTiling: false,
    tileColsLog2: 0,
    tileRowsLog2: 0,
    chromaDeltaQ: false,
    sharpness: 0,
    denoiseLevel: 0,
    tune: AVIFTune.Auto,
    sharpYUV: false,
};
export const mimeType = "image/avif";
export const extension = "avif";
export const bitDepth = [8, 10, 12, 16];
let encoderWASM;
let decoderWASM;
export async function loadEncoder(input) {
    return encoderWASM ??= await loadES(wasmFactoryEnc, input);
}
export async function loadDecoder(input) {
    return decoderWASM ??= await loadES(wasmFactoryDec, input);
}
export function encode(image, options) {
    return encodeES("AVIF Encode", encoderWASM, defaultOptions, image, options);
}
export function decode(input) {
    return check(decoderWASM.decode(input), "AVIF Decode");
}
