// icodec/lib/jpeg.js
import wasmFactoryEnc from "../dist/mozjpeg.js";
import { check, encodeES, loadES } from "./common.js";
export var ColorSpace;
(function (ColorSpace) {
    ColorSpace[ColorSpace["GRAYSCALE"] = 1] = "GRAYSCALE";
    ColorSpace[ColorSpace["RGB"] = 2] = "RGB";
    ColorSpace[ColorSpace["YCbCr"] = 3] = "YCbCr";
})(ColorSpace || (ColorSpace = {}));
// https://github.com/mozilla/mozjpeg/blob/6c9f0897afa1c2738d7222a0a9ab49e8b536a267/jcparam.c#L74
export var Quantization;
(function (Quantization) {
    Quantization[Quantization["JPEG_Annex_K"] = 0] = "JPEG_Annex_K";
    Quantization[Quantization["Flat"] = 1] = "Flat";
    Quantization[Quantization["MSSIM_Tuned_Kodak"] = 2] = "MSSIM_Tuned_Kodak";
    Quantization[Quantization["ImageMagick"] = 3] = "ImageMagick";
    Quantization[Quantization["PSNR_HVS_M_Tuned_Kodak"] = 4] = "PSNR_HVS_M_Tuned_Kodak";
    Quantization[Quantization["Klein_Silverstein_Carney"] = 5] = "Klein_Silverstein_Carney";
    Quantization[Quantization["Watson_Taylor_Borthwick"] = 6] = "Watson_Taylor_Borthwick";
    Quantization[Quantization["Ahumada_Watson_Peterson"] = 7] = "Ahumada_Watson_Peterson";
    Quantization[Quantization["Peterson_Ahumada_Watson"] = 8] = "Peterson_Ahumada_Watson";
})(Quantization || (Quantization = {}));
export const defaultOptions = {
    quality: 75,
    baseline: false,
    arithmetic: false,
    progressive: true,
    optimizeCoding: true,
    smoothing: 0,
    colorSpace: ColorSpace.YCbCr,
    quantTable: Quantization.ImageMagick,
    trellisMultipass: false,
    trellisOptZero: false,
    trellisOptTable: false,
    trellisLoops: 1,
    autoSubsample: true,
    chromaSubsample: 2,
    separateChromaQuality: false,
    chromaQuality: 75,
};
export const bitDepth = [8];
export const mimeType = "image/jpeg";
export const extension = "jpg";
let codecWASM;
export async function loadEncoder(input) {
    return codecWASM = await loadES(wasmFactoryEnc, input);
}
export const loadDecoder = loadEncoder;
export function encode(image, options) {
    return encodeES("JPEG Encode", codecWASM, defaultOptions, image, options);
}
export function decode(input) {
    return check(codecWASM.decode(input), "JPEG Decode");
}
