// icodec/lib/jxl.js
import wasmFactoryEnc from "../dist/jxl-enc.js";
import wasmFactoryDec from "../dist/jxl-dec.js";
import { check, encodeES, loadES } from "./common.js";
// Tristate bool value, `Default` means encoder chooses.
export var Override;
(function (Override) {
    Override[Override["Default"] = -1] = "Default";
    Override[Override["False"] = 0] = "False";
    Override[Override["True"] = 1] = "True";
})(Override || (Override = {}));
export var Predictor;
(function (Predictor) {
    Predictor[Predictor["Default"] = -1] = "Default";
    Predictor[Predictor["Zero"] = 0] = "Zero";
    Predictor[Predictor["Left"] = 1] = "Left";
    Predictor[Predictor["Top"] = 2] = "Top";
    Predictor[Predictor["Average0"] = 3] = "Average0";
    Predictor[Predictor["Select"] = 4] = "Select";
    Predictor[Predictor["Gradient"] = 5] = "Gradient";
    Predictor[Predictor["Weighted"] = 6] = "Weighted";
    Predictor[Predictor["TopRight"] = 7] = "TopRight";
    Predictor[Predictor["TopLeft"] = 8] = "TopLeft";
    Predictor[Predictor["LeftLeft"] = 9] = "LeftLeft";
    Predictor[Predictor["Average1"] = 10] = "Average1";
    Predictor[Predictor["Average2"] = 11] = "Average2";
    Predictor[Predictor["Average3"] = 12] = "Average3";
    Predictor[Predictor["Average4"] = 13] = "Average4";
    // The following predictors are encoder-only.
    Predictor[Predictor["Best"] = 14] = "Best";
    Predictor[Predictor["Variable"] = 15] = "Variable";
})(Predictor || (Predictor = {}));
export const defaultOptions = {
    lossless: false,
    quality: 75,
    alphaQuality: 100,
    effort: 7,
    brotliEffort: -1,
    epf: -1,
    gaborish: -1,
    responsive: -1,
    progressiveDC: -1,
    progressiveAC: -1,
    qProgressiveAC: -1,
    decodingSpeed: 0,
    photonNoiseIso: 0,
    modular: false,
    lossyPalette: false,
    paletteColors: -1,
    iterations: -1,
    modularColorspace: -1,
    modularPredictor: Predictor.Default,
};
export const mimeType = "image/jxl";
export const extension = "jxl";
export const bitDepth = [8, 9, 10, 11, 12, 13, 14, 15, 16];
let encoderWASM;
let decoderWASM;
export async function loadEncoder(input) {
    return encoderWASM ??= await loadES(wasmFactoryEnc, input);
}
export async function loadDecoder(input) {
    return decoderWASM ??= await loadES(wasmFactoryDec, input);
}
export function encode(image, options) {
    return encodeES("JXL Encode", encoderWASM, defaultOptions, image, options);
}
export function decode(input) {
    return check(decoderWASM.decode(input), "JXL Decode");
}
