// icodec/lib/png.js
import wasmFactory, { optimize, png_to_rgba, quantize } from "../dist/pngquant.js";
import { toBitDepth } from "./common.js";
export const defaultOptions = {
    speed: 4,
    quality: 75,
    colors: 256,
    dithering: 1,
    level: 3,
    interlace: false,
    quantize: true,
    bit_depth: 8,
};
export const bitDepth = [8, 16];
export const mimeType = "image/png";
export const extension = "png";
export const loadEncoder = (module_or_path) => wasmFactory({ module_or_path });
export const loadDecoder = loadEncoder;
/**
 * Reduces the colors used in the image at a slight loss, using a combination
 * of vector quantization algorithms.
 *
 * Can be used before other compression algorithm to boost compression ratio.
 */
export function reduceColors(image, options) {
    options = { ...defaultOptions, ...options };
    const { data, width, height } = toBitDepth(image, 8);
    return quantize(data, width, height, { ...defaultOptions, ...options });
}
export function encode(image, options) {
    options = { ...defaultOptions, ...options };
    if (options.quantize) {
        image = toBitDepth(image, 8);
    }
    const { data, width, height, depth } = image;
    options.bit_depth = depth;
    return optimize(data, width, height, { ...defaultOptions, ...options });
}
export function decode(input) {
    const [data, width, depth] = png_to_rgba(input);
    let height = data.byteLength / width / 4;
    if (depth === 16) {
        height /= 2;
    }
    return _icodec_ImageData(data, width, height, depth);
}
