// icodec/lib/webp.js
import wasmFactoryEnc from "../dist/webp-enc.js";
import wasmFactoryDec from "../dist/webp-dec.js";
import { check, encodeES, loadES } from "./common.js";
export var Preprocess;
(function (Preprocess) {
    Preprocess[Preprocess["None"] = 0] = "None";
    Preprocess[Preprocess["SegmentSmooth"] = 1] = "SegmentSmooth";
    Preprocess[Preprocess["Dithering"] = 2] = "Dithering";
})(Preprocess || (Preprocess = {}));
export var AlphaFiltering;
(function (AlphaFiltering) {
    AlphaFiltering[AlphaFiltering["None"] = 0] = "None";
    AlphaFiltering[AlphaFiltering["Fast"] = 1] = "Fast";
    AlphaFiltering[AlphaFiltering["Best"] = 2] = "Best";
})(AlphaFiltering || (AlphaFiltering = {}));
export const defaultOptions = {
    lossless: false,
    nearLossless: 100,
    quality: 75,
    targetSize: 0,
    targetPSNR: 0,
    method: 4,
    snsStrength: 50,
    filterStrength: 60,
    filterSharpness: 0,
    filterType: true,
    segments: 4,
    pass: 1,
    sharpYUV: false,
    preprocessing: Preprocess.None,
    autofilter: false,
    partitionLimit: 0,
    alphaCompression: 1,
    alphaFiltering: AlphaFiltering.Fast,
    alphaQuality: 100,
    exact: false,
    emulateJpegSize: false,
    lowMemory: false,
    // Undocumented options, only for compatibility.
    partitions: 0,
    showCompressed: 0,
    imageHint: 0,
    threadLevel: 0,
    useDeltaPalette: 0,
};
export const bitDepth = [8];
export const mimeType = "image/webp";
export const extension = "webp";
let encoderWASM;
let decoderWASM;
export async function loadEncoder(input) {
    return encoderWASM ??= await loadES(wasmFactoryEnc, input);
}
export async function loadDecoder(input) {
    return decoderWASM ??= await loadES(wasmFactoryDec, input);
}
export function encode(image, options) {
    return encodeES("Webp Encode", encoderWASM, defaultOptions, image, options);
}
export function decode(input) {
    return check(decoderWASM.decode(input), "Webp Decode");
}
