// icodec/lib/wp2.js
import { check, encodeES, loadES } from "./common.js";
import wasmFactoryEnc from "../dist/wp2-enc.js";
import wasmFactoryDec from "../dist/wp2-dec.js";
export var UVMode;
(function (UVMode) {
    UVMode[UVMode["UVAdapt"] = 0] = "UVAdapt";
    UVMode[UVMode["UV420"] = 1] = "UV420";
    UVMode[UVMode["UV444"] = 2] = "UV444";
    UVMode[UVMode["UVAuto"] = 3] = "UVAuto";
})(UVMode || (UVMode = {}));
export var Csp;
(function (Csp) {
    Csp[Csp["YCoCg"] = 0] = "YCoCg";
    Csp[Csp["YCbCr"] = 1] = "YCbCr";
    Csp[Csp["Custom"] = 2] = "Custom";
    Csp[Csp["YIQ"] = 3] = "YIQ";
})(Csp || (Csp = {}));
export const defaultOptions = {
    quality: 75,
    alphaQuality: 100,
    effort: 5,
    pass: 1,
    sns: 50,
    uvMode: UVMode.UVAuto,
    cspType: Csp.YCoCg,
    errorDiffusion: 0,
    useRandomMatrix: false,
};
export const bitDepth = [8];
// WebP 2 will not be released as an image format, but wee need define these properties.
export const mimeType = "image/webp2";
export const extension = "wp2";
let encoderWASM;
let decoderWASM;
export async function loadEncoder(input) {
    return encoderWASM ??= await loadES(wasmFactoryEnc, input);
}
export async function loadDecoder(input) {
    return decoderWASM ??= await loadES(wasmFactoryDec, input);
}
export function encode(image, options) {
    return encodeES("Webp2 Encode", encoderWASM, defaultOptions, image, options);
}
export function decode(input) {
    return check(decoderWASM.decode(input), "Webp2 Decode");
}
