// @casabio/jxl-native/src/index.ts
import { createRequire } from "node:module";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

export class CapabilityMissing extends Error {
  readonly code = "CapabilityMissing";
  readonly cause?: unknown;

  constructor(message: string, cause?: unknown) {
    super(message);
    this.name = "CapabilityMissing";
    if (cause !== undefined) {
      this.cause = cause;
    }
  }
}

export interface NativeBinding {
  version(): string;
  probe(): {
    loaded: boolean;
    path: string;
  };
  createDecoder?: (options: DecoderOptions) => NativeDecoder;
  createEncoder?: (options: EncoderOptions) => NativeEncoder;
}

export interface NativeLoaderOptions {
  prebuiltPath?: string;
  sourcePath?: string;
}

export type PixelFormat = "rgba8" | "rgba16" | "rgbaf32";
export type DecodeStage = "header" | "dc" | "pass" | "final";
export type Region = { x: number; y: number; w: number; h: number };

export interface ExtraChannelDescriptor {
  readonly type: string;
  readonly bitsPerSample: number;
  readonly name: string;
}

export interface DecodedExtraChannel extends ExtraChannelDescriptor {
  readonly pixels: ArrayBuffer;
  readonly pixelFormat: PixelFormat;
}

export interface ImageInfo {
  width: number;
  height: number;
  bitsPerSample: 8 | 16 | 32;
  hasAlpha: boolean;
  hasAnimation: boolean;
  jpegReconstructionAvailable: boolean;
  extraChannels?: readonly ExtraChannelDescriptor[];
}

export type DecodeEvent =
  | { type: "header"; info: ImageInfo }
  | { type: "progress"; stage: DecodeStage; info: ImageInfo; pixels: ArrayBuffer | Uint8Array; format: PixelFormat; region?: Region; pixelStride: number; frameIndex?: number; frameDuration?: number; frameName?: string; isLastFrame?: boolean; animTicksPerSecond?: number; animLoopCount?: number; extraPlanes?: readonly ArrayBuffer[]; extraChannelDescriptors?: readonly DecodedExtraChannel[]; gainMap?: { data: ArrayBuffer } }
  | { type: "final"; info: ImageInfo; pixels: ArrayBuffer | Uint8Array; format: PixelFormat; region?: Region; pixelStride: number; frameIndex?: number; frameDuration?: number; frameName?: string; isLastFrame?: boolean; animTicksPerSecond?: number; animLoopCount?: number; extraPlanes?: readonly ArrayBuffer[]; extraChannelDescriptors?: readonly DecodedExtraChannel[]; gainMap?: { data: ArrayBuffer } }
  | { type: "budget_exceeded"; stage: DecodeStage; info: ImageInfo; pixels: ArrayBuffer | Uint8Array; format: PixelFormat; pixelStride: number }
  | { type: "error"; code: string; message: string };

export interface DecoderOptions {
  format: PixelFormat;
  region: Region | null;
  downsample: 1 | 2 | 4 | 8;
  progressionTarget: "header" | "dc" | "pass" | "final";
  emitEveryPass: boolean;
  preserveIcc: boolean;
  preserveMetadata: boolean;
}

export interface MetadataBoxSpec {
  /** 4-character JXL box type (e.g. "uuid", "xml "). Padded with spaces if shorter. */
  type: string;
  data: Uint8Array;
  /** Compress this box with Brotli. Default false. */
  compress?: boolean;
}

export interface MetadataOptions {
  /** Include ICC profile (default true when iccProfile is non-null). */
  includeICC?: boolean;
  /** Include EXIF box (default true when exif is non-null). */
  includeExif?: boolean;
  /** Include XMP box (default true when xmp is non-null). */
  includeXMP?: boolean;
  /** Compress all metadata boxes with Brotli. Default false. */
  compressBoxes?: boolean;
  /** Force JXL container format even when no metadata boxes are present. */
  forceContainer?: boolean;
  /** Emit raw codestream only — no container, no boxes. Overrides forceContainer. */
  rawCodestream?: boolean;
}

/** Options for attaching an HDR gain map (ISO 21496-1 / JXL jhgm box). */
export interface GainMapOptions {
  /** Pre-encoded JXL naked codestream for the gain map image. */
  data: Uint8Array | ArrayBuffer;
}

/** Sub-settings for Modular encoding mode. Applied alongside or instead of the flat `modular` flag. */
export interface ModularOptions {
  /** 0 = auto group size (libjxl default), positive = explicit group size (power-of-two). */
  groupSize?: number;
  /** Predictor selection (0–15). Major quality/speed tradeoff knob. */
  predictor?: number;
  /** Number of previous channels to use for prediction. */
  nbPrevChannels?: number;
  /** Number of palette colors. 0 = disable palette, -1 = libjxl default. */
  paletteColors?: number;
  /** Allow lossy palette. */
  lossyPalette?: boolean;
  /** Tree learning percent (0–100). -1 = libjxl default. */
  maTreeLearningPercent?: number;
}

/** Descriptor for one extra channel beyond the main color channels. */
export interface ExtraChannel {
  /** Channel type. 'other' maps to JXL_CHANNEL_UNKNOWN (15). */
  type: "alpha" | "depth" | "spot" | "selection" | "other";
  /** Bits per sample for this channel (typically 8, 16, or 32). */
  bitsPerSample: number;
  /** Per-channel encode distance. 0 = lossless; omit to inherit main distance. */
  distance?: number;
  /** Optional human-readable label embedded in the JXL bitstream. */
  name?: string;
}

/** Descriptor for one frame in an animation sequence. */
export interface AnimationFrame {
  data: Uint8Array | ArrayBuffer;
  width: number;
  height: number;
  /** Duration in ticks (see AnimationOptions.ticksPerSecond). */
  duration: number;
  name?: string;
}

/** Animation header options. */
export interface AnimationOptions {
  ticksPerSecond?: number;
  loopCount?: number;
}

export interface EncoderOptions {
  format: PixelFormat;
  width: number;
  height: number;
  hasAlpha: boolean;
  iccProfile: ArrayBuffer | null;
  exif: ArrayBuffer | null;
  xmp: ArrayBuffer | null;
  distance: number | null;
  quality: number | null;
  effort: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;
  /** Brotli effort for aux data (metadata, ICC, EXIF, extra channels). -1 = libjxl default, 0-11 explicit. */
  brotliEffort?: number;
  /** Decoder speed tier hint (0-4). Tells the encoder to structure the codestream for faster decoding. 0 = default, 4 = fastest decode. */
  decodingSpeed?: number;
  /** Target ISO for libjxl synthetic photon noise. 0 or omitted disables it. */
  photonNoiseIso?: number;
  /** Encoder-native downsampling factor before JXL transform/coding. */
  resampling?: 1 | 2 | 4 | 8;
  /** -1 = libjxl auto (default), 0 = VarDCT (lossy), 1 = Modular. */
  modular?: -1 | 0 | 1;
  /**
   * Raw JXL_ENC_FRAME_SETTING_* escape hatch for advanced/experimental features
   * (e.g. patches=8, splines=9, modular predictor, etc.).
   * Applied after all named settings; later entries override earlier ones.
   */
  advancedFrameSettings?: readonly { id: number; value: number }[];
  /** Fine-grained Modular mode sub-settings. Applied after `modular` force flag. */
  modularOptions?: ModularOptions;
  /** Attach an HDR gain map (ISO 21496-1) as a jhgm box. data is a JXL naked codestream. */
  gainMap?: GainMapOptions | null;
  progressive: boolean;
  previewFirst: boolean;
  chunked: boolean;
  /** Container format and per-box options. */
  metadata?: MetadataOptions;
  /** Additional custom metadata boxes to embed. Note: not yet implemented in native binding. */
  customBoxes?: readonly MetadataBoxSpec[];
  /** Animation header options. */
  animation?: AnimationOptions;
  /** Frame data for animation encode. When present, replaces single-image pushPixels. */
  frames?: readonly AnimationFrame[];
  /**
   * Per-channel distance for the alpha channel. 0 = lossless; omit to inherit main distance.
   * Only applied when hasAlpha is true.
   */
  alphaDistance?: number;
  /** Extra channels beyond alpha (e.g. depth, selection mask). */
  extraChannels?: readonly ExtraChannel[];
  /**
   * Pixel data for each extra channel declared in extraChannels.
   * Each entry is a single-channel buffer (width × height × bytesPerSample).
   * May be shorter than extraChannels — missing entries leave the channel data unset.
   */
  extraChannelPlanes?: readonly (ArrayBuffer | Uint8Array)[];
}

export interface NativeDecoder {
  push(chunk: ArrayBuffer | Uint8Array): void | Promise<void>;
  close(): void | Promise<void>;
  events(): AsyncIterable<DecodeEvent>;
  cancel(reason?: string): void | Promise<void>;
  dispose(): void | Promise<void>;
}

export interface NativeEncoder {
  pushPixels(chunk: ArrayBuffer | Uint8Array, region?: Region): void | Promise<void>;
  finish(): void | Promise<void>;
  chunks(): AsyncIterable<ArrayBuffer | Uint8Array>;
  cancel(reason?: string): void | Promise<void>;
  dispose(): void | Promise<void>;
}

export interface NativeCodecFacade {
  createDecoder(options: DecoderOptions): NativeDecoder;
  createEncoder(options: EncoderOptions): NativeEncoder;
}

const require = createRequire(String(import.meta.url));
const packageRoot = dirname(fileURLToPath(String(import.meta.url)));

export function loadNativeBinding(options: NativeLoaderOptions = {}): NativeBinding {
  const candidates = [
    options.prebuiltPath ?? resolvePrebuiltBinary(),
    options.sourcePath ?? resolveSourceBinary()
  ];

  let lastError: unknown;
  for (const candidate of candidates) {
    try {
      const binding = require(candidate) as NativeBinding;
      ensureBindingLoaded(binding, candidate);
      return binding;
    } catch (error) {
      lastError = error;
    }
  }

  throw new CapabilityMissing("jxl-native addon unavailable; falling back to WASM is required", lastError);
}

export function createNativeCodecFacade(binding: NativeBinding): NativeCodecFacade {
  if (typeof binding.createDecoder !== "function" || typeof binding.createEncoder !== "function") {
    throw new CapabilityMissing("jxl-native addon does not expose createDecoder/createEncoder");
  }
  ensureBindingLoaded(binding, "native binding");
  return {
    createDecoder(options) {
      return binding.createDecoder!(options);
    },
    createEncoder(options) {
      return binding.createEncoder!(options);
    },
  };
}

export function createDecoder(options: DecoderOptions): NativeDecoder {
  return createNativeCodecFacade(loadNativeBinding()).createDecoder(options);
}

export function createEncoder(options: EncoderOptions): NativeEncoder {
  return createNativeCodecFacade(loadNativeBinding()).createEncoder(options);
}

function resolvePrebuiltBinary(): string {
  const platform = process?.platform ?? "unknown";
  const arch = process?.arch ?? "unknown";
  const base = join(packageRoot, "..", "prebuilds");
  const candidate = resolve(base, `${platform}-${arch}`, "jxl-native.node");
  return candidate;
}

function resolveSourceBinary(): string {
  const release = resolve(packageRoot, "..", "build", "Release", "jxl_native.node");
  const debug = resolve(packageRoot, "..", "build", "Debug", "jxl_native.node");
  return fileExists(release) ? release : fileExists(debug) ? debug : release;
}

function fileExists(path: string): boolean {
  try {
    require("node:fs").accessSync(path);
    return true;
  } catch {
    return false;
  }
}

function ensureBindingLoaded(binding: NativeBinding, label: string): void {
  if (typeof binding.version === "function" && binding.version().includes("scaffold")) {
    throw new CapabilityMissing(`jxl-native addon at ${label} is still the scaffold stub`);
  }
  if (typeof binding.probe !== "function") return;
  const probe = binding.probe();
  if (typeof probe.path === "string" && probe.path.toLowerCase().includes("stub")) {
    throw new CapabilityMissing(`jxl-native addon at ${label} is still the scaffold stub`, probe);
  }
  if (probe.loaded !== true) {
    throw new CapabilityMissing(`jxl-native addon at ${label} is present but not loaded`, probe);
  }
}
